packages/ra-core/src/auth/useCanAccessRecordSources.ts

```diff
@@ -0,0 +1,152 @@
+import {
+    useQueries,
+    UseQueryOptions,
+    UseQueryResult,
+} from '@tanstack/react-query';
+import useAuthProvider from './useAuthProvider';
+import useLogoutIfAccessDenied from './useLogoutIfAccessDenied';
+import { useEvent } from '../util';
+import { useEffect } from 'react';
+import { useResourceContext } from '../core';
+
+export const combine = <ErrorType>(
+    results: UseQueryResult<
+        {
+            canAccess: boolean;
+            source: string;
+        },
+        ErrorType
+    >[]
+): {
+    data?: Record<string, boolean>;
+    isPending: boolean;
+    isError: boolean;
+    error?: ErrorType;
+} => {
+    return {
+        data: results
+            ? results.reduce(
+                  (acc, { data }) => {
+                      if (!data) {
+                          return acc;
+                      }
+                      const { source, canAccess } = data;
+                      return {
+                          ...acc,
+                          [source]: canAccess,
+                      };
+                  },
+                  {} as Record<string, boolean>
+              )
+            : undefined,
+        isPending: results.some(result => result.isPending),
+        isError: results.some(result => result.isError),
+        error: results.find(result => result.error)?.error || undefined,
+    };
+};
+
+/**
+ * Checks if the user can access a resource keys.
+ *
+ * `useCanAccessRecordSources` returns an object describing the state of the RBAC request.
+ * As calls to the `authProvider` are asynchronous, the hook returns
+ * a `loading` state in addition to the `canAccess` key.
+ *
+ * @example
+ * import { useCanAccessRecordSources } from 'react-admin';
+ *
+ * const DeleteUserButton = ({ record }) => {
+ *     const { isPending, canAccess } = useCanAccessRecordSources([{ action: 'delete', resource: 'users', record, sources: ['id', 'name', 'email'] }]);
+ *
+ *     if (isPending) {
+ *         return null;
+ *     }
+ *     return (
+ *         <SimpleList
+ *              primaryText={record => canAccess.name ? record.name : ''}
+ *              secondaryText={record => canAccess.email ? record.email : ''}
+ *              tertiaryText={record => canAccess.id ? record.id : ''}
+ *          />
+ *     );
+ * };
+ */
+export const useCanAccessRecordSources = <ErrorType extends Error = Error>(
```

Comment:
I think we should extract the generic parallel `canAccess` fetcher based on `useQueries`, keeping only the logic of the ${resource}.${source}` combination here.